// mongodb/lib/core/sdam/dist/common.dev.js
'use strict'; // shared state names

var STATE_CLOSING = 'closing';
var STATE_CLOSED = 'closed';
var STATE_CONNECTING = 'connecting';
var STATE_CONNECTED = 'connected'; // An enumeration of topology types we know about

var TopologyType = {
  Single: 'Single',
  ReplicaSetNoPrimary: 'ReplicaSetNoPrimary',
  ReplicaSetWithPrimary: 'ReplicaSetWithPrimary',
  Sharded: 'Sharded',
  Unknown: 'Unknown'
}; // An enumeration of server types we know about

var ServerType = {
  Standalone: 'Standalone',
  Mongos: 'Mongos',
  PossiblePrimary: 'PossiblePrimary',
  RSPrimary: 'RSPrimary',
  RSSecondary: 'RSSecondary',
  RSArbiter: 'RSArbiter',
  RSOther: 'RSOther',
  RSGhost: 'RSGhost',
  Unknown: 'Unknown'
}; // helper to get a server's type that works for both legacy and unified topologies

function serverType(server) {
  var description = server.s.description || server.s.serverDescription;
  if (description.topologyType === TopologyType.Single) return description.servers[0].type;
  return description.type;
}

var TOPOLOGY_DEFAULTS = {
  useUnifiedTopology: true,
  localThresholdMS: 15,
  serverSelectionTimeoutMS: 30000,
  heartbeatFrequencyMS: 10000,
  minHeartbeatFrequencyMS: 500
};

function drainTimerQueue(queue) {
  queue.forEach(clearTimeout);
  queue.clear();
}

function clearAndRemoveTimerFrom(timer, timers) {
  clearTimeout(timer);
  return timers["delete"](timer);
}

module.exports = {
  STATE_CLOSING: STATE_CLOSING,
  STATE_CLOSED: STATE_CLOSED,
  STATE_CONNECTING: STATE_CONNECTING,
  STATE_CONNECTED: STATE_CONNECTED,
  TOPOLOGY_DEFAULTS: TOPOLOGY_DEFAULTS,
  TopologyType: TopologyType,
  ServerType: ServerType,
  serverType: serverType,
  drainTimerQueue: drainTimerQueue,
  clearAndRemoveTimerFrom: clearAndRemoveTimerFrom
};
