// googleapis/build/src/apis/genomics/dist/v2alpha1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.genomics_v2alpha1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var genomics_v2alpha1;

(function (genomics_v2alpha1) {
  /**
   * Genomics API
   *
   * Uploads, processes, queries, and searches Genomics data in the cloud.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const genomics = google.genomics('v2alpha1');
   * ```
   */
  var Genomics = function Genomics(options, google) {
    _classCallCheck(this, Genomics);

    this.context = {
      _options: options || {},
      google: google
    };
    this.pipelines = new Resource$Pipelines(this.context);
    this.projects = new Resource$Projects(this.context);
    this.workers = new Resource$Workers(this.context);
  };

  genomics_v2alpha1.Genomics = Genomics;

  var Resource$Pipelines =
  /*#__PURE__*/
  function () {
    function Resource$Pipelines(context) {
      _classCallCheck(this, Resource$Pipelines);

      this.context = context;
    }

    _createClass(Resource$Pipelines, [{
      key: "run",
      value: function run(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://genomics.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v2alpha1/pipelines:run').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Pipelines;
  }();

  genomics_v2alpha1.Resource$Pipelines = Resource$Pipelines;

  var Resource$Projects = function Resource$Projects(context) {
    _classCallCheck(this, Resource$Projects);

    this.context = context;
    this.operations = new Resource$Projects$Operations(this.context);
    this.workers = new Resource$Projects$Workers(this.context);
  };

  genomics_v2alpha1.Resource$Projects = Resource$Projects;

  var Resource$Projects$Operations =
  /*#__PURE__*/
  function () {
    function Resource$Projects$Operations(context) {
      _classCallCheck(this, Resource$Projects$Operations);

      this.context = context;
    }

    _createClass(Resource$Projects$Operations, [{
      key: "cancel",
      value: function cancel(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://genomics.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v2alpha1/{+name}:cancel').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "get",
      value: function get(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://genomics.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v2alpha1/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "list",
      value: function list(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://genomics.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v2alpha1/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Projects$Operations;
  }();

  genomics_v2alpha1.Resource$Projects$Operations = Resource$Projects$Operations;

  var Resource$Projects$Workers =
  /*#__PURE__*/
  function () {
    function Resource$Projects$Workers(context) {
      _classCallCheck(this, Resource$Projects$Workers);

      this.context = context;
    }

    _createClass(Resource$Projects$Workers, [{
      key: "checkIn",
      value: function checkIn(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://genomics.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v2alpha1/{+id}:checkIn').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['id'],
          pathParams: ['id'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Projects$Workers;
  }();

  genomics_v2alpha1.Resource$Projects$Workers = Resource$Projects$Workers;

  var Resource$Workers =
  /*#__PURE__*/
  function () {
    function Resource$Workers(context) {
      _classCallCheck(this, Resource$Workers);

      this.context = context;
    }

    _createClass(Resource$Workers, [{
      key: "checkIn",
      value: function checkIn(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://genomics.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v2alpha1/workers/{id}:checkIn').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['id'],
          pathParams: ['id'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Workers;
  }();

  genomics_v2alpha1.Resource$Workers = Resource$Workers;
})(genomics_v2alpha1 = exports.genomics_v2alpha1 || (exports.genomics_v2alpha1 = {}));
