// googleapis/build/src/apis/serviceconsumermanagement/dist/v1beta1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.serviceconsumermanagement_v1beta1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var serviceconsumermanagement_v1beta1;

(function (serviceconsumermanagement_v1beta1) {
  /**
   * Service Consumer Management API
   *
   * Manages the service consumers of a Service Infrastructure service.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const serviceconsumermanagement = google.serviceconsumermanagement('v1beta1');
   * ```
   */
  var Serviceconsumermanagement = function Serviceconsumermanagement(options, google) {
    _classCallCheck(this, Serviceconsumermanagement);

    this.context = {
      _options: options || {},
      google: google
    };
    this.operations = new Resource$Operations(this.context);
    this.services = new Resource$Services(this.context);
  };

  serviceconsumermanagement_v1beta1.Serviceconsumermanagement = Serviceconsumermanagement;

  var Resource$Operations =
  /*#__PURE__*/
  function () {
    function Resource$Operations(context) {
      _classCallCheck(this, Resource$Operations);

      this.context = context;
    }

    _createClass(Resource$Operations, [{
      key: "get",
      value: function get(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://serviceconsumermanagement.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Operations;
  }();

  serviceconsumermanagement_v1beta1.Resource$Operations = Resource$Operations;

  var Resource$Services = function Resource$Services(context) {
    _classCallCheck(this, Resource$Services);

    this.context = context;
    this.consumerQuotaMetrics = new Resource$Services$Consumerquotametrics(this.context);
  };

  serviceconsumermanagement_v1beta1.Resource$Services = Resource$Services;

  var Resource$Services$Consumerquotametrics =
  /*#__PURE__*/
  function () {
    function Resource$Services$Consumerquotametrics(context) {
      _classCallCheck(this, Resource$Services$Consumerquotametrics);

      this.context = context;
      this.limits = new Resource$Services$Consumerquotametrics$Limits(this.context);
    }

    _createClass(Resource$Services$Consumerquotametrics, [{
      key: "get",
      value: function get(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://serviceconsumermanagement.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "importProducerOverrides",
      value: function importProducerOverrides(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://serviceconsumermanagement.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+parent}/consumerQuotaMetrics:importProducerOverrides').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "list",
      value: function list(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://serviceconsumermanagement.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+parent}/consumerQuotaMetrics').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Services$Consumerquotametrics;
  }();

  serviceconsumermanagement_v1beta1.Resource$Services$Consumerquotametrics = Resource$Services$Consumerquotametrics;

  var Resource$Services$Consumerquotametrics$Limits =
  /*#__PURE__*/
  function () {
    function Resource$Services$Consumerquotametrics$Limits(context) {
      _classCallCheck(this, Resource$Services$Consumerquotametrics$Limits);

      this.context = context;
      this.producerOverrides = new Resource$Services$Consumerquotametrics$Limits$Produceroverrides(this.context);
    }

    _createClass(Resource$Services$Consumerquotametrics$Limits, [{
      key: "get",
      value: function get(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://serviceconsumermanagement.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Services$Consumerquotametrics$Limits;
  }();

  serviceconsumermanagement_v1beta1.Resource$Services$Consumerquotametrics$Limits = Resource$Services$Consumerquotametrics$Limits;

  var Resource$Services$Consumerquotametrics$Limits$Produceroverrides =
  /*#__PURE__*/
  function () {
    function Resource$Services$Consumerquotametrics$Limits$Produceroverrides(context) {
      _classCallCheck(this, Resource$Services$Consumerquotametrics$Limits$Produceroverrides);

      this.context = context;
    }

    _createClass(Resource$Services$Consumerquotametrics$Limits$Produceroverrides, [{
      key: "create",
      value: function create(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://serviceconsumermanagement.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+parent}/producerOverrides').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "delete",
      value: function _delete(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://serviceconsumermanagement.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'DELETE'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "list",
      value: function list(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://serviceconsumermanagement.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+parent}/producerOverrides').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "patch",
      value: function patch(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://serviceconsumermanagement.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'PATCH'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Services$Consumerquotametrics$Limits$Produceroverrides;
  }();

  serviceconsumermanagement_v1beta1.Resource$Services$Consumerquotametrics$Limits$Produceroverrides = Resource$Services$Consumerquotametrics$Limits$Produceroverrides;
})(serviceconsumermanagement_v1beta1 = exports.serviceconsumermanagement_v1beta1 || (exports.serviceconsumermanagement_v1beta1 = {}));
