// googleapis/build/src/apis/driveactivity/dist/v2.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.driveactivity_v2 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var driveactivity_v2;

(function (driveactivity_v2) {
  /**
   * Drive Activity API
   *
   * Provides a historical view of activity in Google Drive.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const driveactivity = google.driveactivity('v2');
   * ```
   */
  var Driveactivity = function Driveactivity(options, google) {
    _classCallCheck(this, Driveactivity);

    this.context = {
      _options: options || {},
      google: google
    };
    this.activity = new Resource$Activity(this.context);
  };

  driveactivity_v2.Driveactivity = Driveactivity;

  var Resource$Activity =
  /*#__PURE__*/
  function () {
    function Resource$Activity(context) {
      _classCallCheck(this, Resource$Activity);

      this.context = context;
    }

    _createClass(Resource$Activity, [{
      key: "query",
      value: function query(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://driveactivity.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v2/activity:query').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Activity;
  }();

  driveactivity_v2.Resource$Activity = Resource$Activity;
})(driveactivity_v2 = exports.driveactivity_v2 || (exports.driveactivity_v2 = {}));
