// uri-js/dist/esnext/dist/util.dev.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.merge = merge;
exports.subexp = subexp;
exports.typeOf = typeOf;
exports.toUpperCase = toUpperCase;
exports.toArray = toArray;
exports.assign = assign;

function merge() {
  for (var _len = arguments.length, sets = new Array(_len), _key = 0; _key < _len; _key++) {
    sets[_key] = arguments[_key];
  }

  if (sets.length > 1) {
    sets[0] = sets[0].slice(0, -1);
    var xl = sets.length - 1;

    for (var x = 1; x < xl; ++x) {
      sets[x] = sets[x].slice(1, -1);
    }

    sets[xl] = sets[xl].slice(1);
    return sets.join('');
  } else {
    return sets[0];
  }
}

function subexp(str) {
  return "(?:" + str + ")";
}

function typeOf(o) {
  return o === undefined ? "undefined" : o === null ? "null" : Object.prototype.toString.call(o).split(" ").pop().split("]").shift().toLowerCase();
}

function toUpperCase(str) {
  return str.toUpperCase();
}

function toArray(obj) {
  return obj !== undefined && obj !== null ? obj instanceof Array ? obj : typeof obj.length !== "number" || obj.split || obj.setInterval || obj.call ? [obj] : Array.prototype.slice.call(obj) : [];
}

function assign(target, source) {
  var obj = target;

  if (source) {
    for (var key in source) {
      obj[key] = source[key];
    }
  }

  return obj;
}
