// googleapis/build/src/apis/playcustomapp/dist/v1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.playcustomapp_v1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var playcustomapp_v1;

(function (playcustomapp_v1) {
  /**
   * Google Play Custom App Publishing API
   *
   * API to create and publish custom Android apps
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const playcustomapp = google.playcustomapp('v1');
   * ```
   */
  var Playcustomapp = function Playcustomapp(options, google) {
    _classCallCheck(this, Playcustomapp);

    this.context = {
      _options: options || {},
      google: google
    };
    this.accounts = new Resource$Accounts(this.context);
  };

  playcustomapp_v1.Playcustomapp = Playcustomapp;

  var Resource$Accounts = function Resource$Accounts(context) {
    _classCallCheck(this, Resource$Accounts);

    this.context = context;
    this.customApps = new Resource$Accounts$Customapps(this.context);
  };

  playcustomapp_v1.Resource$Accounts = Resource$Accounts;

  var Resource$Accounts$Customapps =
  /*#__PURE__*/
  function () {
    function Resource$Accounts$Customapps(context) {
      _classCallCheck(this, Resource$Accounts$Customapps);

      this.context = context;
    }

    _createClass(Resource$Accounts$Customapps, [{
      key: "create",
      value: function create(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://playcustomapp.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/playcustomapp/v1/accounts/{account}/customApps').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          mediaUrl: (rootUrl + '/upload/playcustomapp/v1/accounts/{account}/customApps').replace(/([^:]\/)\/+/g, '$1'),
          requiredParams: ['account'],
          pathParams: ['account'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Accounts$Customapps;
  }();

  playcustomapp_v1.Resource$Accounts$Customapps = Resource$Accounts$Customapps;
})(playcustomapp_v1 = exports.playcustomapp_v1 || (exports.playcustomapp_v1 = {}));
