// googleapis/build/src/apis/displayvideo/dist/v1beta2.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.displayvideo_v1beta2 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var displayvideo_v1beta2;

(function (displayvideo_v1beta2) {
  /**
   * Display &amp; Video 360 API
   *
   * Display &amp; Video 360 API allows users to manage and create campaigns and reports.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const displayvideo = google.displayvideo('v1beta2');
   * ```
   */
  var Displayvideo = function Displayvideo(options, google) {
    _classCallCheck(this, Displayvideo);

    this.context = {
      _options: options || {},
      google: google
    };
    this.media = new Resource$Media(this.context);
    this.sdfdownloadtasks = new Resource$Sdfdownloadtasks(this.context);
  };

  displayvideo_v1beta2.Displayvideo = Displayvideo;

  var Resource$Media =
  /*#__PURE__*/
  function () {
    function Resource$Media(context) {
      _classCallCheck(this, Resource$Media);

      this.context = context;
    }

    _createClass(Resource$Media, [{
      key: "download",
      value: function download(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://displayvideo.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/download/{+resourceName}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['resourceName'],
          pathParams: ['resourceName'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Media;
  }();

  displayvideo_v1beta2.Resource$Media = Resource$Media;

  var Resource$Sdfdownloadtasks = function Resource$Sdfdownloadtasks(context) {
    _classCallCheck(this, Resource$Sdfdownloadtasks);

    this.context = context;
    this.operations = new Resource$Sdfdownloadtasks$Operations(this.context);
  };

  displayvideo_v1beta2.Resource$Sdfdownloadtasks = Resource$Sdfdownloadtasks;

  var Resource$Sdfdownloadtasks$Operations =
  /*#__PURE__*/
  function () {
    function Resource$Sdfdownloadtasks$Operations(context) {
      _classCallCheck(this, Resource$Sdfdownloadtasks$Operations);

      this.context = context;
    }

    _createClass(Resource$Sdfdownloadtasks$Operations, [{
      key: "get",
      value: function get(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://displayvideo.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta2/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Sdfdownloadtasks$Operations;
  }();

  displayvideo_v1beta2.Resource$Sdfdownloadtasks$Operations = Resource$Sdfdownloadtasks$Operations;
})(displayvideo_v1beta2 = exports.displayvideo_v1beta2 || (exports.displayvideo_v1beta2 = {}));
