// googleapis/build/src/apis/customsearch/dist/v1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.customsearch_v1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var customsearch_v1;

(function (customsearch_v1) {
  /**
   * Custom Search API
   *
   * Searches over a website or collection of websites
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const customsearch = google.customsearch('v1');
   * ```
   */
  var Customsearch = function Customsearch(options, google) {
    _classCallCheck(this, Customsearch);

    this.context = {
      _options: options || {},
      google: google
    };
    this.cse = new Resource$Cse(this.context);
  };

  customsearch_v1.Customsearch = Customsearch;

  var Resource$Cse =
  /*#__PURE__*/
  function () {
    function Resource$Cse(context) {
      _classCallCheck(this, Resource$Cse);

      this.context = context;
      this.siterestrict = new Resource$Cse$Siterestrict(this.context);
    }

    _createClass(Resource$Cse, [{
      key: "list",
      value: function list(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://customsearch.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/customsearch/v1').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Cse;
  }();

  customsearch_v1.Resource$Cse = Resource$Cse;

  var Resource$Cse$Siterestrict =
  /*#__PURE__*/
  function () {
    function Resource$Cse$Siterestrict(context) {
      _classCallCheck(this, Resource$Cse$Siterestrict);

      this.context = context;
    }

    _createClass(Resource$Cse$Siterestrict, [{
      key: "list",
      value: function list(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://customsearch.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/customsearch/v1/siterestrict').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Cse$Siterestrict;
  }();

  customsearch_v1.Resource$Cse$Siterestrict = Resource$Cse$Siterestrict;
})(customsearch_v1 = exports.customsearch_v1 || (exports.customsearch_v1 = {}));
