// pg-protocol/src/testing/dist/buffer-list.js
"use strict";
exports.__esModule = true;
var BufferList = /** @class */ (function () {
    function BufferList(buffers) {
        if (buffers === void 0) { buffers = []; }
        this.buffers = buffers;
    }
    BufferList.prototype.add = function (buffer, front) {
        this.buffers[front ? 'unshift' : 'push'](buffer);
        return this;
    };
    BufferList.prototype.addInt16 = function (val, front) {
        return this.add(Buffer.from([val >>> 8, val >>> 0]), front);
    };
    BufferList.prototype.getByteLength = function (initial) {
        return this.buffers.reduce(function (previous, current) {
            return previous + current.length;
        }, initial || 0);
    };
    BufferList.prototype.addInt32 = function (val, first) {
        return this.add(Buffer.from([(val >>> 24) & 0xff, (val >>> 16) & 0xff, (val >>> 8) & 0xff, (val >>> 0) & 0xff]), first);
    };
    BufferList.prototype.addCString = function (val, front) {
        var len = Buffer.byteLength(val);
        var buffer = Buffer.alloc(len + 1);
        buffer.write(val);
        buffer[len] = 0;
        return this.add(buffer, front);
    };
    BufferList.prototype.addString = function (val, front) {
        var len = Buffer.byteLength(val);
        var buffer = Buffer.alloc(len);
        buffer.write(val);
        return this.add(buffer, front);
    };
    BufferList.prototype.addChar = function (char, first) {
        return this.add(Buffer.from(char, 'utf8'), first);
    };
    BufferList.prototype.addByte = function (byte) {
        return this.add(Buffer.from([byte]));
    };
    BufferList.prototype.join = function (appendLength, char) {
        var length = this.getByteLength();
        if (appendLength) {
            this.addInt32(length + 4, true);
            return this.join(false, char);
        }
        if (char) {
            this.addChar(char, true);
            length++;
        }
        var result = Buffer.alloc(length);
        var index = 0;
        this.buffers.forEach(function (buffer) {
            buffer.copy(result, index, 0);
            index += buffer.length;
        });
        return result;
    };
    BufferList.concat = function () {
        var total = new BufferList();
        for (var i = 0; i < arguments.length; i++) {
            total.add(arguments[i]);
        }
        return total.join();
    };
    return BufferList;
}());
exports["default"] = BufferList;
