// googleapis/build/src/apis/vision/dist/v1p1beta1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.vision_v1p1beta1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var vision_v1p1beta1;

(function (vision_v1p1beta1) {
  /**
   * Cloud Vision API
   *
   * Integrates Google Vision features, including image labeling, face, logo, and landmark detection, optical character recognition (OCR), and detection of explicit content, into applications.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const vision = google.vision('v1p1beta1');
   * ```
   */
  var Vision = function Vision(options, google) {
    _classCallCheck(this, Vision);

    this.context = {
      _options: options || {},
      google: google
    };
    this.files = new Resource$Files(this.context);
    this.images = new Resource$Images(this.context);
    this.projects = new Resource$Projects(this.context);
  };

  vision_v1p1beta1.Vision = Vision;

  var Resource$Files =
  /*#__PURE__*/
  function () {
    function Resource$Files(context) {
      _classCallCheck(this, Resource$Files);

      this.context = context;
    }

    _createClass(Resource$Files, [{
      key: "annotate",
      value: function annotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/files:annotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "asyncBatchAnnotate",
      value: function asyncBatchAnnotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/files:asyncBatchAnnotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Files;
  }();

  vision_v1p1beta1.Resource$Files = Resource$Files;

  var Resource$Images =
  /*#__PURE__*/
  function () {
    function Resource$Images(context) {
      _classCallCheck(this, Resource$Images);

      this.context = context;
    }

    _createClass(Resource$Images, [{
      key: "annotate",
      value: function annotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/images:annotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "asyncBatchAnnotate",
      value: function asyncBatchAnnotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/images:asyncBatchAnnotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Images;
  }();

  vision_v1p1beta1.Resource$Images = Resource$Images;

  var Resource$Projects = function Resource$Projects(context) {
    _classCallCheck(this, Resource$Projects);

    this.context = context;
    this.files = new Resource$Projects$Files(this.context);
    this.images = new Resource$Projects$Images(this.context);
    this.locations = new Resource$Projects$Locations(this.context);
  };

  vision_v1p1beta1.Resource$Projects = Resource$Projects;

  var Resource$Projects$Files =
  /*#__PURE__*/
  function () {
    function Resource$Projects$Files(context) {
      _classCallCheck(this, Resource$Projects$Files);

      this.context = context;
    }

    _createClass(Resource$Projects$Files, [{
      key: "annotate",
      value: function annotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/{+parent}/files:annotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "asyncBatchAnnotate",
      value: function asyncBatchAnnotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/{+parent}/files:asyncBatchAnnotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Projects$Files;
  }();

  vision_v1p1beta1.Resource$Projects$Files = Resource$Projects$Files;

  var Resource$Projects$Images =
  /*#__PURE__*/
  function () {
    function Resource$Projects$Images(context) {
      _classCallCheck(this, Resource$Projects$Images);

      this.context = context;
    }

    _createClass(Resource$Projects$Images, [{
      key: "annotate",
      value: function annotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/{+parent}/images:annotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "asyncBatchAnnotate",
      value: function asyncBatchAnnotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/{+parent}/images:asyncBatchAnnotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Projects$Images;
  }();

  vision_v1p1beta1.Resource$Projects$Images = Resource$Projects$Images;

  var Resource$Projects$Locations = function Resource$Projects$Locations(context) {
    _classCallCheck(this, Resource$Projects$Locations);

    this.context = context;
    this.files = new Resource$Projects$Locations$Files(this.context);
    this.images = new Resource$Projects$Locations$Images(this.context);
  };

  vision_v1p1beta1.Resource$Projects$Locations = Resource$Projects$Locations;

  var Resource$Projects$Locations$Files =
  /*#__PURE__*/
  function () {
    function Resource$Projects$Locations$Files(context) {
      _classCallCheck(this, Resource$Projects$Locations$Files);

      this.context = context;
    }

    _createClass(Resource$Projects$Locations$Files, [{
      key: "annotate",
      value: function annotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/{+parent}/files:annotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "asyncBatchAnnotate",
      value: function asyncBatchAnnotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/{+parent}/files:asyncBatchAnnotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Projects$Locations$Files;
  }();

  vision_v1p1beta1.Resource$Projects$Locations$Files = Resource$Projects$Locations$Files;

  var Resource$Projects$Locations$Images =
  /*#__PURE__*/
  function () {
    function Resource$Projects$Locations$Images(context) {
      _classCallCheck(this, Resource$Projects$Locations$Images);

      this.context = context;
    }

    _createClass(Resource$Projects$Locations$Images, [{
      key: "annotate",
      value: function annotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/{+parent}/images:annotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "asyncBatchAnnotate",
      value: function asyncBatchAnnotate(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://vision.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1p1beta1/{+parent}/images:asyncBatchAnnotate').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Projects$Locations$Images;
  }();

  vision_v1p1beta1.Resource$Projects$Locations$Images = Resource$Projects$Locations$Images;
})(vision_v1p1beta1 = exports.vision_v1p1beta1 || (exports.vision_v1p1beta1 = {}));
