// pg-promise/lib/dist/special-query.dev.js
"use strict";

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

/*
 * Copyright (c) 2015-present, Vitaly Tomilov
 *
 * See the LICENSE file at the top-level directory of this distribution
 * for licensing information.
 *
 * Removal or modification of this copyright notice is prohibited.
 */
var specialQueryType = {
  result: 0,
  multiResult: 1,
  stream: 2
};

var SpecialQuery = function SpecialQuery(type) {
  _classCallCheck(this, SpecialQuery);

  this.isResult = type === specialQueryType.result; // type used implicitly

  this.isStream = type === specialQueryType.stream;
  this.isMultiResult = type === specialQueryType.multiResult;
};

var cache = {
  resultQuery: new SpecialQuery(specialQueryType.result),
  multiResultQuery: new SpecialQuery(specialQueryType.multiResult),
  streamQuery: new SpecialQuery(specialQueryType.stream)
};
module.exports = Object.assign({
  SpecialQuery: SpecialQuery
}, cache);
