// googleapis/build/src/apis/localservices/dist/v1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.localservices_v1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var localservices_v1;

(function (localservices_v1) {
  /**
   * Local Services API
   *
   *
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const localservices = google.localservices('v1');
   * ```
   */
  var Localservices = function Localservices(options, google) {
    _classCallCheck(this, Localservices);

    this.context = {
      _options: options || {},
      google: google
    };
    this.accountReports = new Resource$Accountreports(this.context);
    this.detailedLeadReports = new Resource$Detailedleadreports(this.context);
  };

  localservices_v1.Localservices = Localservices;

  var Resource$Accountreports =
  /*#__PURE__*/
  function () {
    function Resource$Accountreports(context) {
      _classCallCheck(this, Resource$Accountreports);

      this.context = context;
    }

    _createClass(Resource$Accountreports, [{
      key: "search",
      value: function search(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://localservices.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/accountReports:search').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Accountreports;
  }();

  localservices_v1.Resource$Accountreports = Resource$Accountreports;

  var Resource$Detailedleadreports =
  /*#__PURE__*/
  function () {
    function Resource$Detailedleadreports(context) {
      _classCallCheck(this, Resource$Detailedleadreports);

      this.context = context;
    }

    _createClass(Resource$Detailedleadreports, [{
      key: "search",
      value: function search(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://localservices.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/detailedLeadReports:search').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Detailedleadreports;
  }();

  localservices_v1.Resource$Detailedleadreports = Resource$Detailedleadreports;
})(localservices_v1 = exports.localservices_v1 || (exports.localservices_v1 = {}));
