// googleapis/build/src/apis/dist/index.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/*! THIS FILE IS AUTO-GENERATED */

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GeneratedAPIs = exports.APIS = void 0;

var abusiveexperiencereport_1 = require("./abusiveexperiencereport");

var acceleratedmobilepageurl_1 = require("./acceleratedmobilepageurl");

var accessapproval_1 = require("./accessapproval");

var accesscontextmanager_1 = require("./accesscontextmanager");

var adexchangebuyer_1 = require("./adexchangebuyer");

var adexchangebuyer2_1 = require("./adexchangebuyer2");

var adexperiencereport_1 = require("./adexperiencereport");

var admin_1 = require("./admin");

var admob_1 = require("./admob");

var adsense_1 = require("./adsense");

var adsensehost_1 = require("./adsensehost");

var alertcenter_1 = require("./alertcenter");

var analytics_1 = require("./analytics");

var analyticsadmin_1 = require("./analyticsadmin");

var analyticsdata_1 = require("./analyticsdata");

var analyticsreporting_1 = require("./analyticsreporting");

var androiddeviceprovisioning_1 = require("./androiddeviceprovisioning");

var androidenterprise_1 = require("./androidenterprise");

var androidmanagement_1 = require("./androidmanagement");

var androidpublisher_1 = require("./androidpublisher");

var apigateway_1 = require("./apigateway");

var apikeys_1 = require("./apikeys");

var appengine_1 = require("./appengine");

var appsactivity_1 = require("./appsactivity");

var area120tables_1 = require("./area120tables");

var artifactregistry_1 = require("./artifactregistry");

var assuredworkloads_1 = require("./assuredworkloads");

var bigquery_1 = require("./bigquery");

var bigqueryconnection_1 = require("./bigqueryconnection");

var bigquerydatatransfer_1 = require("./bigquerydatatransfer");

var bigqueryreservation_1 = require("./bigqueryreservation");

var bigtableadmin_1 = require("./bigtableadmin");

var billingbudgets_1 = require("./billingbudgets");

var binaryauthorization_1 = require("./binaryauthorization");

var blogger_1 = require("./blogger");

var books_1 = require("./books");

var calendar_1 = require("./calendar");

var chat_1 = require("./chat");

var chromemanagement_1 = require("./chromemanagement");

var chromepolicy_1 = require("./chromepolicy");

var chromeuxreport_1 = require("./chromeuxreport");

var civicinfo_1 = require("./civicinfo");

var classroom_1 = require("./classroom");

var cloudasset_1 = require("./cloudasset");

var cloudbilling_1 = require("./cloudbilling");

var cloudbuild_1 = require("./cloudbuild");

var cloudchannel_1 = require("./cloudchannel");

var clouddebugger_1 = require("./clouddebugger");

var clouderrorreporting_1 = require("./clouderrorreporting");

var cloudfunctions_1 = require("./cloudfunctions");

var cloudidentity_1 = require("./cloudidentity");

var cloudiot_1 = require("./cloudiot");

var cloudkms_1 = require("./cloudkms");

var cloudprofiler_1 = require("./cloudprofiler");

var cloudresourcemanager_1 = require("./cloudresourcemanager");

var cloudscheduler_1 = require("./cloudscheduler");

var cloudsearch_1 = require("./cloudsearch");

var cloudshell_1 = require("./cloudshell");

var cloudtasks_1 = require("./cloudtasks");

var cloudtrace_1 = require("./cloudtrace");

var composer_1 = require("./composer");

var compute_1 = require("./compute");

var container_1 = require("./container");

var containeranalysis_1 = require("./containeranalysis");

var content_1 = require("./content");

var customsearch_1 = require("./customsearch");

var datacatalog_1 = require("./datacatalog");

var dataflow_1 = require("./dataflow");

var datafusion_1 = require("./datafusion");

var datalabeling_1 = require("./datalabeling");

var datamigration_1 = require("./datamigration");

var dataproc_1 = require("./dataproc");

var datastore_1 = require("./datastore");

var deploymentmanager_1 = require("./deploymentmanager");

var dfareporting_1 = require("./dfareporting");

var dialogflow_1 = require("./dialogflow");

var digitalassetlinks_1 = require("./digitalassetlinks");

var discovery_1 = require("./discovery");

var displayvideo_1 = require("./displayvideo");

var dlp_1 = require("./dlp");

var dns_1 = require("./dns");

var docs_1 = require("./docs");

var documentai_1 = require("./documentai");

var domains_1 = require("./domains");

var domainsrdap_1 = require("./domainsrdap");

var doubleclickbidmanager_1 = require("./doubleclickbidmanager");

var doubleclicksearch_1 = require("./doubleclicksearch");

var drive_1 = require("./drive");

var driveactivity_1 = require("./driveactivity");

var essentialcontacts_1 = require("./essentialcontacts");

var eventarc_1 = require("./eventarc");

var factchecktools_1 = require("./factchecktools");

var fcm_1 = require("./fcm");

var file_1 = require("./file");

var firebase_1 = require("./firebase");

var firebasedatabase_1 = require("./firebasedatabase");

var firebasedynamiclinks_1 = require("./firebasedynamiclinks");

var firebasehosting_1 = require("./firebasehosting");

var firebaseml_1 = require("./firebaseml");

var firebaserules_1 = require("./firebaserules");

var firebasestorage_1 = require("./firebasestorage");

var firestore_1 = require("./firestore");

var fitness_1 = require("./fitness");

var games_1 = require("./games");

var gamesConfiguration_1 = require("./gamesConfiguration");

var gamesManagement_1 = require("./gamesManagement");

var gameservices_1 = require("./gameservices");

var genomics_1 = require("./genomics");

var gkehub_1 = require("./gkehub");

var gmail_1 = require("./gmail");

var gmailpostmastertools_1 = require("./gmailpostmastertools");

var groupsmigration_1 = require("./groupsmigration");

var groupssettings_1 = require("./groupssettings");

var healthcare_1 = require("./healthcare");

var homegraph_1 = require("./homegraph");

var iam_1 = require("./iam");

var iamcredentials_1 = require("./iamcredentials");

var iap_1 = require("./iap");

var identitytoolkit_1 = require("./identitytoolkit");

var indexing_1 = require("./indexing");

var jobs_1 = require("./jobs");

var kgsearch_1 = require("./kgsearch");

var language_1 = require("./language");

var libraryagent_1 = require("./libraryagent");

var licensing_1 = require("./licensing");

var lifesciences_1 = require("./lifesciences");

var localservices_1 = require("./localservices");

var logging_1 = require("./logging");

var managedidentities_1 = require("./managedidentities");

var manufacturers_1 = require("./manufacturers");

var memcache_1 = require("./memcache");

var metastore_1 = require("./metastore");

var ml_1 = require("./ml");

var monitoring_1 = require("./monitoring");

var mybusinessaccountmanagement_1 = require("./mybusinessaccountmanagement");

var mybusinesslodging_1 = require("./mybusinesslodging");

var networkconnectivity_1 = require("./networkconnectivity");

var networkmanagement_1 = require("./networkmanagement");

var notebooks_1 = require("./notebooks");

var oauth2_1 = require("./oauth2");

var ondemandscanning_1 = require("./ondemandscanning");

var orgpolicy_1 = require("./orgpolicy");

var osconfig_1 = require("./osconfig");

var oslogin_1 = require("./oslogin");

var pagespeedonline_1 = require("./pagespeedonline");

var people_1 = require("./people");

var playablelocations_1 = require("./playablelocations");

var playcustomapp_1 = require("./playcustomapp");

var plus_1 = require("./plus");

var policysimulator_1 = require("./policysimulator");

var policytroubleshooter_1 = require("./policytroubleshooter");

var poly_1 = require("./poly");

var privateca_1 = require("./privateca");

var prod_tt_sasportal_1 = require("./prod_tt_sasportal");

var pubsub_1 = require("./pubsub");

var pubsublite_1 = require("./pubsublite");

var realtimebidding_1 = require("./realtimebidding");

var recommendationengine_1 = require("./recommendationengine");

var recommender_1 = require("./recommender");

var redis_1 = require("./redis");

var remotebuildexecution_1 = require("./remotebuildexecution");

var reseller_1 = require("./reseller");

var retail_1 = require("./retail");

var run_1 = require("./run");

var runtimeconfig_1 = require("./runtimeconfig");

var safebrowsing_1 = require("./safebrowsing");

var sasportal_1 = require("./sasportal");

var script_1 = require("./script");

var searchconsole_1 = require("./searchconsole");

var secretmanager_1 = require("./secretmanager");

var securitycenter_1 = require("./securitycenter");

var serviceconsumermanagement_1 = require("./serviceconsumermanagement");

var servicecontrol_1 = require("./servicecontrol");

var servicedirectory_1 = require("./servicedirectory");

var servicemanagement_1 = require("./servicemanagement");

var servicenetworking_1 = require("./servicenetworking");

var serviceusage_1 = require("./serviceusage");

var sheets_1 = require("./sheets");

var siteVerification_1 = require("./siteVerification");

var slides_1 = require("./slides");

var smartdevicemanagement_1 = require("./smartdevicemanagement");

var sourcerepo_1 = require("./sourcerepo");

var spanner_1 = require("./spanner");

var speech_1 = require("./speech");

var sql_1 = require("./sql");

var sqladmin_1 = require("./sqladmin");

var storage_1 = require("./storage");

var storagetransfer_1 = require("./storagetransfer");

var streetviewpublish_1 = require("./streetviewpublish");

var sts_1 = require("./sts");

var tagmanager_1 = require("./tagmanager");

var tasks_1 = require("./tasks");

var testing_1 = require("./testing");

var texttospeech_1 = require("./texttospeech");

var toolresults_1 = require("./toolresults");

var tpu_1 = require("./tpu");

var trafficdirector_1 = require("./trafficdirector");

var transcoder_1 = require("./transcoder");

var translate_1 = require("./translate");

var vault_1 = require("./vault");

var vectortile_1 = require("./vectortile");

var verifiedaccess_1 = require("./verifiedaccess");

var videointelligence_1 = require("./videointelligence");

var vision_1 = require("./vision");

var webfonts_1 = require("./webfonts");

var webmasters_1 = require("./webmasters");

var webrisk_1 = require("./webrisk");

var websecurityscanner_1 = require("./websecurityscanner");

var workflowexecutions_1 = require("./workflowexecutions");

var workflows_1 = require("./workflows");

var youtube_1 = require("./youtube");

var youtubeAnalytics_1 = require("./youtubeAnalytics");

var youtubereporting_1 = require("./youtubereporting");

exports.APIS = {
  abusiveexperiencereport: abusiveexperiencereport_1.VERSIONS,
  acceleratedmobilepageurl: acceleratedmobilepageurl_1.VERSIONS,
  accessapproval: accessapproval_1.VERSIONS,
  accesscontextmanager: accesscontextmanager_1.VERSIONS,
  adexchangebuyer: adexchangebuyer_1.VERSIONS,
  adexchangebuyer2: adexchangebuyer2_1.VERSIONS,
  adexperiencereport: adexperiencereport_1.VERSIONS,
  admin: admin_1.VERSIONS,
  admob: admob_1.VERSIONS,
  adsense: adsense_1.VERSIONS,
  adsensehost: adsensehost_1.VERSIONS,
  alertcenter: alertcenter_1.VERSIONS,
  analytics: analytics_1.VERSIONS,
  analyticsadmin: analyticsadmin_1.VERSIONS,
  analyticsdata: analyticsdata_1.VERSIONS,
  analyticsreporting: analyticsreporting_1.VERSIONS,
  androiddeviceprovisioning: androiddeviceprovisioning_1.VERSIONS,
  androidenterprise: androidenterprise_1.VERSIONS,
  androidmanagement: androidmanagement_1.VERSIONS,
  androidpublisher: androidpublisher_1.VERSIONS,
  apigateway: apigateway_1.VERSIONS,
  apikeys: apikeys_1.VERSIONS,
  appengine: appengine_1.VERSIONS,
  appsactivity: appsactivity_1.VERSIONS,
  area120tables: area120tables_1.VERSIONS,
  artifactregistry: artifactregistry_1.VERSIONS,
  assuredworkloads: assuredworkloads_1.VERSIONS,
  bigquery: bigquery_1.VERSIONS,
  bigqueryconnection: bigqueryconnection_1.VERSIONS,
  bigquerydatatransfer: bigquerydatatransfer_1.VERSIONS,
  bigqueryreservation: bigqueryreservation_1.VERSIONS,
  bigtableadmin: bigtableadmin_1.VERSIONS,
  billingbudgets: billingbudgets_1.VERSIONS,
  binaryauthorization: binaryauthorization_1.VERSIONS,
  blogger: blogger_1.VERSIONS,
  books: books_1.VERSIONS,
  calendar: calendar_1.VERSIONS,
  chat: chat_1.VERSIONS,
  chromemanagement: chromemanagement_1.VERSIONS,
  chromepolicy: chromepolicy_1.VERSIONS,
  chromeuxreport: chromeuxreport_1.VERSIONS,
  civicinfo: civicinfo_1.VERSIONS,
  classroom: classroom_1.VERSIONS,
  cloudasset: cloudasset_1.VERSIONS,
  cloudbilling: cloudbilling_1.VERSIONS,
  cloudbuild: cloudbuild_1.VERSIONS,
  cloudchannel: cloudchannel_1.VERSIONS,
  clouddebugger: clouddebugger_1.VERSIONS,
  clouderrorreporting: clouderrorreporting_1.VERSIONS,
  cloudfunctions: cloudfunctions_1.VERSIONS,
  cloudidentity: cloudidentity_1.VERSIONS,
  cloudiot: cloudiot_1.VERSIONS,
  cloudkms: cloudkms_1.VERSIONS,
  cloudprofiler: cloudprofiler_1.VERSIONS,
  cloudresourcemanager: cloudresourcemanager_1.VERSIONS,
  cloudscheduler: cloudscheduler_1.VERSIONS,
  cloudsearch: cloudsearch_1.VERSIONS,
  cloudshell: cloudshell_1.VERSIONS,
  cloudtasks: cloudtasks_1.VERSIONS,
  cloudtrace: cloudtrace_1.VERSIONS,
  composer: composer_1.VERSIONS,
  compute: compute_1.VERSIONS,
  container: container_1.VERSIONS,
  containeranalysis: containeranalysis_1.VERSIONS,
  content: content_1.VERSIONS,
  customsearch: customsearch_1.VERSIONS,
  datacatalog: datacatalog_1.VERSIONS,
  dataflow: dataflow_1.VERSIONS,
  datafusion: datafusion_1.VERSIONS,
  datalabeling: datalabeling_1.VERSIONS,
  datamigration: datamigration_1.VERSIONS,
  dataproc: dataproc_1.VERSIONS,
  datastore: datastore_1.VERSIONS,
  deploymentmanager: deploymentmanager_1.VERSIONS,
  dfareporting: dfareporting_1.VERSIONS,
  dialogflow: dialogflow_1.VERSIONS,
  digitalassetlinks: digitalassetlinks_1.VERSIONS,
  discovery: discovery_1.VERSIONS,
  displayvideo: displayvideo_1.VERSIONS,
  dlp: dlp_1.VERSIONS,
  dns: dns_1.VERSIONS,
  docs: docs_1.VERSIONS,
  documentai: documentai_1.VERSIONS,
  domains: domains_1.VERSIONS,
  domainsrdap: domainsrdap_1.VERSIONS,
  doubleclickbidmanager: doubleclickbidmanager_1.VERSIONS,
  doubleclicksearch: doubleclicksearch_1.VERSIONS,
  drive: drive_1.VERSIONS,
  driveactivity: driveactivity_1.VERSIONS,
  essentialcontacts: essentialcontacts_1.VERSIONS,
  eventarc: eventarc_1.VERSIONS,
  factchecktools: factchecktools_1.VERSIONS,
  fcm: fcm_1.VERSIONS,
  file: file_1.VERSIONS,
  firebase: firebase_1.VERSIONS,
  firebasedatabase: firebasedatabase_1.VERSIONS,
  firebasedynamiclinks: firebasedynamiclinks_1.VERSIONS,
  firebasehosting: firebasehosting_1.VERSIONS,
  firebaseml: firebaseml_1.VERSIONS,
  firebaserules: firebaserules_1.VERSIONS,
  firebasestorage: firebasestorage_1.VERSIONS,
  firestore: firestore_1.VERSIONS,
  fitness: fitness_1.VERSIONS,
  games: games_1.VERSIONS,
  gamesConfiguration: gamesConfiguration_1.VERSIONS,
  gamesManagement: gamesManagement_1.VERSIONS,
  gameservices: gameservices_1.VERSIONS,
  genomics: genomics_1.VERSIONS,
  gkehub: gkehub_1.VERSIONS,
  gmail: gmail_1.VERSIONS,
  gmailpostmastertools: gmailpostmastertools_1.VERSIONS,
  groupsmigration: groupsmigration_1.VERSIONS,
  groupssettings: groupssettings_1.VERSIONS,
  healthcare: healthcare_1.VERSIONS,
  homegraph: homegraph_1.VERSIONS,
  iam: iam_1.VERSIONS,
  iamcredentials: iamcredentials_1.VERSIONS,
  iap: iap_1.VERSIONS,
  identitytoolkit: identitytoolkit_1.VERSIONS,
  indexing: indexing_1.VERSIONS,
  jobs: jobs_1.VERSIONS,
  kgsearch: kgsearch_1.VERSIONS,
  language: language_1.VERSIONS,
  libraryagent: libraryagent_1.VERSIONS,
  licensing: licensing_1.VERSIONS,
  lifesciences: lifesciences_1.VERSIONS,
  localservices: localservices_1.VERSIONS,
  logging: logging_1.VERSIONS,
  managedidentities: managedidentities_1.VERSIONS,
  manufacturers: manufacturers_1.VERSIONS,
  memcache: memcache_1.VERSIONS,
  metastore: metastore_1.VERSIONS,
  ml: ml_1.VERSIONS,
  monitoring: monitoring_1.VERSIONS,
  mybusinessaccountmanagement: mybusinessaccountmanagement_1.VERSIONS,
  mybusinesslodging: mybusinesslodging_1.VERSIONS,
  networkconnectivity: networkconnectivity_1.VERSIONS,
  networkmanagement: networkmanagement_1.VERSIONS,
  notebooks: notebooks_1.VERSIONS,
  oauth2: oauth2_1.VERSIONS,
  ondemandscanning: ondemandscanning_1.VERSIONS,
  orgpolicy: orgpolicy_1.VERSIONS,
  osconfig: osconfig_1.VERSIONS,
  oslogin: oslogin_1.VERSIONS,
  pagespeedonline: pagespeedonline_1.VERSIONS,
  people: people_1.VERSIONS,
  playablelocations: playablelocations_1.VERSIONS,
  playcustomapp: playcustomapp_1.VERSIONS,
  plus: plus_1.VERSIONS,
  policysimulator: policysimulator_1.VERSIONS,
  policytroubleshooter: policytroubleshooter_1.VERSIONS,
  poly: poly_1.VERSIONS,
  privateca: privateca_1.VERSIONS,
  prod_tt_sasportal: prod_tt_sasportal_1.VERSIONS,
  pubsub: pubsub_1.VERSIONS,
  pubsublite: pubsublite_1.VERSIONS,
  realtimebidding: realtimebidding_1.VERSIONS,
  recommendationengine: recommendationengine_1.VERSIONS,
  recommender: recommender_1.VERSIONS,
  redis: redis_1.VERSIONS,
  remotebuildexecution: remotebuildexecution_1.VERSIONS,
  reseller: reseller_1.VERSIONS,
  retail: retail_1.VERSIONS,
  run: run_1.VERSIONS,
  runtimeconfig: runtimeconfig_1.VERSIONS,
  safebrowsing: safebrowsing_1.VERSIONS,
  sasportal: sasportal_1.VERSIONS,
  script: script_1.VERSIONS,
  searchconsole: searchconsole_1.VERSIONS,
  secretmanager: secretmanager_1.VERSIONS,
  securitycenter: securitycenter_1.VERSIONS,
  serviceconsumermanagement: serviceconsumermanagement_1.VERSIONS,
  servicecontrol: servicecontrol_1.VERSIONS,
  servicedirectory: servicedirectory_1.VERSIONS,
  servicemanagement: servicemanagement_1.VERSIONS,
  servicenetworking: servicenetworking_1.VERSIONS,
  serviceusage: serviceusage_1.VERSIONS,
  sheets: sheets_1.VERSIONS,
  siteVerification: siteVerification_1.VERSIONS,
  slides: slides_1.VERSIONS,
  smartdevicemanagement: smartdevicemanagement_1.VERSIONS,
  sourcerepo: sourcerepo_1.VERSIONS,
  spanner: spanner_1.VERSIONS,
  speech: speech_1.VERSIONS,
  sql: sql_1.VERSIONS,
  sqladmin: sqladmin_1.VERSIONS,
  storage: storage_1.VERSIONS,
  storagetransfer: storagetransfer_1.VERSIONS,
  streetviewpublish: streetviewpublish_1.VERSIONS,
  sts: sts_1.VERSIONS,
  tagmanager: tagmanager_1.VERSIONS,
  tasks: tasks_1.VERSIONS,
  testing: testing_1.VERSIONS,
  texttospeech: texttospeech_1.VERSIONS,
  toolresults: toolresults_1.VERSIONS,
  tpu: tpu_1.VERSIONS,
  trafficdirector: trafficdirector_1.VERSIONS,
  transcoder: transcoder_1.VERSIONS,
  translate: translate_1.VERSIONS,
  vault: vault_1.VERSIONS,
  vectortile: vectortile_1.VERSIONS,
  verifiedaccess: verifiedaccess_1.VERSIONS,
  videointelligence: videointelligence_1.VERSIONS,
  vision: vision_1.VERSIONS,
  webfonts: webfonts_1.VERSIONS,
  webmasters: webmasters_1.VERSIONS,
  webrisk: webrisk_1.VERSIONS,
  websecurityscanner: websecurityscanner_1.VERSIONS,
  workflowexecutions: workflowexecutions_1.VERSIONS,
  workflows: workflows_1.VERSIONS,
  youtube: youtube_1.VERSIONS,
  youtubeAnalytics: youtubeAnalytics_1.VERSIONS,
  youtubereporting: youtubereporting_1.VERSIONS
};

var GeneratedAPIs = function GeneratedAPIs() {
  _classCallCheck(this, GeneratedAPIs);

  this.abusiveexperiencereport = abusiveexperiencereport_1.abusiveexperiencereport;
  this.acceleratedmobilepageurl = acceleratedmobilepageurl_1.acceleratedmobilepageurl;
  this.accessapproval = accessapproval_1.accessapproval;
  this.accesscontextmanager = accesscontextmanager_1.accesscontextmanager;
  this.adexchangebuyer = adexchangebuyer_1.adexchangebuyer;
  this.adexchangebuyer2 = adexchangebuyer2_1.adexchangebuyer2;
  this.adexperiencereport = adexperiencereport_1.adexperiencereport;
  this.admin = admin_1.admin;
  this.admob = admob_1.admob;
  this.adsense = adsense_1.adsense;
  this.adsensehost = adsensehost_1.adsensehost;
  this.alertcenter = alertcenter_1.alertcenter;
  this.analytics = analytics_1.analytics;
  this.analyticsadmin = analyticsadmin_1.analyticsadmin;
  this.analyticsdata = analyticsdata_1.analyticsdata;
  this.analyticsreporting = analyticsreporting_1.analyticsreporting;
  this.androiddeviceprovisioning = androiddeviceprovisioning_1.androiddeviceprovisioning;
  this.androidenterprise = androidenterprise_1.androidenterprise;
  this.androidmanagement = androidmanagement_1.androidmanagement;
  this.androidpublisher = androidpublisher_1.androidpublisher;
  this.apigateway = apigateway_1.apigateway;
  this.apikeys = apikeys_1.apikeys;
  this.appengine = appengine_1.appengine;
  this.appsactivity = appsactivity_1.appsactivity;
  this.area120tables = area120tables_1.area120tables;
  this.artifactregistry = artifactregistry_1.artifactregistry;
  this.assuredworkloads = assuredworkloads_1.assuredworkloads;
  this.bigquery = bigquery_1.bigquery;
  this.bigqueryconnection = bigqueryconnection_1.bigqueryconnection;
  this.bigquerydatatransfer = bigquerydatatransfer_1.bigquerydatatransfer;
  this.bigqueryreservation = bigqueryreservation_1.bigqueryreservation;
  this.bigtableadmin = bigtableadmin_1.bigtableadmin;
  this.billingbudgets = billingbudgets_1.billingbudgets;
  this.binaryauthorization = binaryauthorization_1.binaryauthorization;
  this.blogger = blogger_1.blogger;
  this.books = books_1.books;
  this.calendar = calendar_1.calendar;
  this.chat = chat_1.chat;
  this.chromemanagement = chromemanagement_1.chromemanagement;
  this.chromepolicy = chromepolicy_1.chromepolicy;
  this.chromeuxreport = chromeuxreport_1.chromeuxreport;
  this.civicinfo = civicinfo_1.civicinfo;
  this.classroom = classroom_1.classroom;
  this.cloudasset = cloudasset_1.cloudasset;
  this.cloudbilling = cloudbilling_1.cloudbilling;
  this.cloudbuild = cloudbuild_1.cloudbuild;
  this.cloudchannel = cloudchannel_1.cloudchannel;
  this.clouddebugger = clouddebugger_1.clouddebugger;
  this.clouderrorreporting = clouderrorreporting_1.clouderrorreporting;
  this.cloudfunctions = cloudfunctions_1.cloudfunctions;
  this.cloudidentity = cloudidentity_1.cloudidentity;
  this.cloudiot = cloudiot_1.cloudiot;
  this.cloudkms = cloudkms_1.cloudkms;
  this.cloudprofiler = cloudprofiler_1.cloudprofiler;
  this.cloudresourcemanager = cloudresourcemanager_1.cloudresourcemanager;
  this.cloudscheduler = cloudscheduler_1.cloudscheduler;
  this.cloudsearch = cloudsearch_1.cloudsearch;
  this.cloudshell = cloudshell_1.cloudshell;
  this.cloudtasks = cloudtasks_1.cloudtasks;
  this.cloudtrace = cloudtrace_1.cloudtrace;
  this.composer = composer_1.composer;
  this.compute = compute_1.compute;
  this.container = container_1.container;
  this.containeranalysis = containeranalysis_1.containeranalysis;
  this.content = content_1.content;
  this.customsearch = customsearch_1.customsearch;
  this.datacatalog = datacatalog_1.datacatalog;
  this.dataflow = dataflow_1.dataflow;
  this.datafusion = datafusion_1.datafusion;
  this.datalabeling = datalabeling_1.datalabeling;
  this.datamigration = datamigration_1.datamigration;
  this.dataproc = dataproc_1.dataproc;
  this.datastore = datastore_1.datastore;
  this.deploymentmanager = deploymentmanager_1.deploymentmanager;
  this.dfareporting = dfareporting_1.dfareporting;
  this.dialogflow = dialogflow_1.dialogflow;
  this.digitalassetlinks = digitalassetlinks_1.digitalassetlinks;
  this.discovery = discovery_1.discovery;
  this.displayvideo = displayvideo_1.displayvideo;
  this.dlp = dlp_1.dlp;
  this.dns = dns_1.dns;
  this.docs = docs_1.docs;
  this.documentai = documentai_1.documentai;
  this.domains = domains_1.domains;
  this.domainsrdap = domainsrdap_1.domainsrdap;
  this.doubleclickbidmanager = doubleclickbidmanager_1.doubleclickbidmanager;
  this.doubleclicksearch = doubleclicksearch_1.doubleclicksearch;
  this.drive = drive_1.drive;
  this.driveactivity = driveactivity_1.driveactivity;
  this.essentialcontacts = essentialcontacts_1.essentialcontacts;
  this.eventarc = eventarc_1.eventarc;
  this.factchecktools = factchecktools_1.factchecktools;
  this.fcm = fcm_1.fcm;
  this.file = file_1.file;
  this.firebase = firebase_1.firebase;
  this.firebasedatabase = firebasedatabase_1.firebasedatabase;
  this.firebasedynamiclinks = firebasedynamiclinks_1.firebasedynamiclinks;
  this.firebasehosting = firebasehosting_1.firebasehosting;
  this.firebaseml = firebaseml_1.firebaseml;
  this.firebaserules = firebaserules_1.firebaserules;
  this.firebasestorage = firebasestorage_1.firebasestorage;
  this.firestore = firestore_1.firestore;
  this.fitness = fitness_1.fitness;
  this.games = games_1.games;
  this.gamesConfiguration = gamesConfiguration_1.gamesConfiguration;
  this.gamesManagement = gamesManagement_1.gamesManagement;
  this.gameservices = gameservices_1.gameservices;
  this.genomics = genomics_1.genomics;
  this.gkehub = gkehub_1.gkehub;
  this.gmail = gmail_1.gmail;
  this.gmailpostmastertools = gmailpostmastertools_1.gmailpostmastertools;
  this.groupsmigration = groupsmigration_1.groupsmigration;
  this.groupssettings = groupssettings_1.groupssettings;
  this.healthcare = healthcare_1.healthcare;
  this.homegraph = homegraph_1.homegraph;
  this.iam = iam_1.iam;
  this.iamcredentials = iamcredentials_1.iamcredentials;
  this.iap = iap_1.iap;
  this.identitytoolkit = identitytoolkit_1.identitytoolkit;
  this.indexing = indexing_1.indexing;
  this.jobs = jobs_1.jobs;
  this.kgsearch = kgsearch_1.kgsearch;
  this.language = language_1.language;
  this.libraryagent = libraryagent_1.libraryagent;
  this.licensing = licensing_1.licensing;
  this.lifesciences = lifesciences_1.lifesciences;
  this.localservices = localservices_1.localservices;
  this.logging = logging_1.logging;
  this.managedidentities = managedidentities_1.managedidentities;
  this.manufacturers = manufacturers_1.manufacturers;
  this.memcache = memcache_1.memcache;
  this.metastore = metastore_1.metastore;
  this.ml = ml_1.ml;
  this.monitoring = monitoring_1.monitoring;
  this.mybusinessaccountmanagement = mybusinessaccountmanagement_1.mybusinessaccountmanagement;
  this.mybusinesslodging = mybusinesslodging_1.mybusinesslodging;
  this.networkconnectivity = networkconnectivity_1.networkconnectivity;
  this.networkmanagement = networkmanagement_1.networkmanagement;
  this.notebooks = notebooks_1.notebooks;
  this.oauth2 = oauth2_1.oauth2;
  this.ondemandscanning = ondemandscanning_1.ondemandscanning;
  this.orgpolicy = orgpolicy_1.orgpolicy;
  this.osconfig = osconfig_1.osconfig;
  this.oslogin = oslogin_1.oslogin;
  this.pagespeedonline = pagespeedonline_1.pagespeedonline;
  this.people = people_1.people;
  this.playablelocations = playablelocations_1.playablelocations;
  this.playcustomapp = playcustomapp_1.playcustomapp;
  this.plus = plus_1.plus;
  this.policysimulator = policysimulator_1.policysimulator;
  this.policytroubleshooter = policytroubleshooter_1.policytroubleshooter;
  this.poly = poly_1.poly;
  this.privateca = privateca_1.privateca;
  this.prod_tt_sasportal = prod_tt_sasportal_1.prod_tt_sasportal;
  this.pubsub = pubsub_1.pubsub;
  this.pubsublite = pubsublite_1.pubsublite;
  this.realtimebidding = realtimebidding_1.realtimebidding;
  this.recommendationengine = recommendationengine_1.recommendationengine;
  this.recommender = recommender_1.recommender;
  this.redis = redis_1.redis;
  this.remotebuildexecution = remotebuildexecution_1.remotebuildexecution;
  this.reseller = reseller_1.reseller;
  this.retail = retail_1.retail;
  this.run = run_1.run;
  this.runtimeconfig = runtimeconfig_1.runtimeconfig;
  this.safebrowsing = safebrowsing_1.safebrowsing;
  this.sasportal = sasportal_1.sasportal;
  this.script = script_1.script;
  this.searchconsole = searchconsole_1.searchconsole;
  this.secretmanager = secretmanager_1.secretmanager;
  this.securitycenter = securitycenter_1.securitycenter;
  this.serviceconsumermanagement = serviceconsumermanagement_1.serviceconsumermanagement;
  this.servicecontrol = servicecontrol_1.servicecontrol;
  this.servicedirectory = servicedirectory_1.servicedirectory;
  this.servicemanagement = servicemanagement_1.servicemanagement;
  this.servicenetworking = servicenetworking_1.servicenetworking;
  this.serviceusage = serviceusage_1.serviceusage;
  this.sheets = sheets_1.sheets;
  this.siteVerification = siteVerification_1.siteVerification;
  this.slides = slides_1.slides;
  this.smartdevicemanagement = smartdevicemanagement_1.smartdevicemanagement;
  this.sourcerepo = sourcerepo_1.sourcerepo;
  this.spanner = spanner_1.spanner;
  this.speech = speech_1.speech;
  this.sql = sql_1.sql;
  this.sqladmin = sqladmin_1.sqladmin;
  this.storage = storage_1.storage;
  this.storagetransfer = storagetransfer_1.storagetransfer;
  this.streetviewpublish = streetviewpublish_1.streetviewpublish;
  this.sts = sts_1.sts;
  this.tagmanager = tagmanager_1.tagmanager;
  this.tasks = tasks_1.tasks;
  this.testing = testing_1.testing;
  this.texttospeech = texttospeech_1.texttospeech;
  this.toolresults = toolresults_1.toolresults;
  this.tpu = tpu_1.tpu;
  this.trafficdirector = trafficdirector_1.trafficdirector;
  this.transcoder = transcoder_1.transcoder;
  this.translate = translate_1.translate;
  this.vault = vault_1.vault;
  this.vectortile = vectortile_1.vectortile;
  this.verifiedaccess = verifiedaccess_1.verifiedaccess;
  this.videointelligence = videointelligence_1.videointelligence;
  this.vision = vision_1.vision;
  this.webfonts = webfonts_1.webfonts;
  this.webmasters = webmasters_1.webmasters;
  this.webrisk = webrisk_1.webrisk;
  this.websecurityscanner = websecurityscanner_1.websecurityscanner;
  this.workflowexecutions = workflowexecutions_1.workflowexecutions;
  this.workflows = workflows_1.workflows;
  this.youtube = youtube_1.youtube;
  this.youtubeAnalytics = youtubeAnalytics_1.youtubeAnalytics;
  this.youtubereporting = youtubereporting_1.youtubereporting;
};

exports.GeneratedAPIs = GeneratedAPIs;
