// googleapis/build/src/apis/run/dist/v1beta1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.run_v1beta1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var run_v1beta1;

(function (run_v1beta1) {
  /**
   * Cloud Run Admin API
   *
   * Deploy and manage user provided container images that scale automatically based on HTTP traffic.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const run = google.run('v1beta1');
   * ```
   */
  var Run = function Run(options, google) {
    _classCallCheck(this, Run);

    this.context = {
      _options: options || {},
      google: google
    };
    this.customresourcedefinitions = new Resource$Customresourcedefinitions(this.context);
    this.namespaces = new Resource$Namespaces(this.context);
    this.projects = new Resource$Projects(this.context);
  };

  run_v1beta1.Run = Run;

  var Resource$Customresourcedefinitions =
  /*#__PURE__*/
  function () {
    function Resource$Customresourcedefinitions(context) {
      _classCallCheck(this, Resource$Customresourcedefinitions);

      this.context = context;
    }

    _createClass(Resource$Customresourcedefinitions, [{
      key: "list",
      value: function list(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://run.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/apis/apiextensions.k8s.io/v1beta1/customresourcedefinitions').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Customresourcedefinitions;
  }();

  run_v1beta1.Resource$Customresourcedefinitions = Resource$Customresourcedefinitions;

  var Resource$Namespaces = function Resource$Namespaces(context) {
    _classCallCheck(this, Resource$Namespaces);

    this.context = context;
    this.customresourcedefinitions = new Resource$Namespaces$Customresourcedefinitions(this.context);
  };

  run_v1beta1.Resource$Namespaces = Resource$Namespaces;

  var Resource$Namespaces$Customresourcedefinitions =
  /*#__PURE__*/
  function () {
    function Resource$Namespaces$Customresourcedefinitions(context) {
      _classCallCheck(this, Resource$Namespaces$Customresourcedefinitions);

      this.context = context;
    }

    _createClass(Resource$Namespaces$Customresourcedefinitions, [{
      key: "get",
      value: function get(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://run.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/apis/apiextensions.k8s.io/v1beta1/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Namespaces$Customresourcedefinitions;
  }();

  run_v1beta1.Resource$Namespaces$Customresourcedefinitions = Resource$Namespaces$Customresourcedefinitions;

  var Resource$Projects = function Resource$Projects(context) {
    _classCallCheck(this, Resource$Projects);

    this.context = context;
    this.locations = new Resource$Projects$Locations(this.context);
  };

  run_v1beta1.Resource$Projects = Resource$Projects;

  var Resource$Projects$Locations = function Resource$Projects$Locations(context) {
    _classCallCheck(this, Resource$Projects$Locations);

    this.context = context;
    this.customresourcedefinitions = new Resource$Projects$Locations$Customresourcedefinitions(this.context);
  };

  run_v1beta1.Resource$Projects$Locations = Resource$Projects$Locations;

  var Resource$Projects$Locations$Customresourcedefinitions =
  /*#__PURE__*/
  function () {
    function Resource$Projects$Locations$Customresourcedefinitions(context) {
      _classCallCheck(this, Resource$Projects$Locations$Customresourcedefinitions);

      this.context = context;
    }

    _createClass(Resource$Projects$Locations$Customresourcedefinitions, [{
      key: "get",
      value: function get(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://run.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+name}').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['name'],
          pathParams: ['name'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "list",
      value: function list(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://run.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta1/{+parent}/customresourcedefinitions').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['parent'],
          pathParams: ['parent'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Projects$Locations$Customresourcedefinitions;
  }();

  run_v1beta1.Resource$Projects$Locations$Customresourcedefinitions = Resource$Projects$Locations$Customresourcedefinitions;
})(run_v1beta1 = exports.run_v1beta1 || (exports.run_v1beta1 = {}));
