// googleapis/build/src/apis/pagespeedonline/dist/v5.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pagespeedonline_v5 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var pagespeedonline_v5;

(function (pagespeedonline_v5) {
  /**
   * PageSpeed Insights API
   *
   * The PageSpeed Insights API lets you analyze the performance of your website with a simple API. It offers tailored suggestions for how you can optimize your site, and lets you easily integrate PageSpeed Insights analysis into your development tools and workflow.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const pagespeedonline = google.pagespeedonline('v5');
   * ```
   */
  var Pagespeedonline = function Pagespeedonline(options, google) {
    _classCallCheck(this, Pagespeedonline);

    this.context = {
      _options: options || {},
      google: google
    };
    this.pagespeedapi = new Resource$Pagespeedapi(this.context);
  };

  pagespeedonline_v5.Pagespeedonline = Pagespeedonline;

  var Resource$Pagespeedapi =
  /*#__PURE__*/
  function () {
    function Resource$Pagespeedapi(context) {
      _classCallCheck(this, Resource$Pagespeedapi);

      this.context = context;
    }

    _createClass(Resource$Pagespeedapi, [{
      key: "runpagespeed",
      value: function runpagespeed(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://pagespeedonline.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/pagespeedonline/v5/runPagespeed').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['url'],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Pagespeedapi;
  }();

  pagespeedonline_v5.Resource$Pagespeedapi = Resource$Pagespeedapi;
})(pagespeedonline_v5 = exports.pagespeedonline_v5 || (exports.pagespeedonline_v5 = {}));
