// googleapis/build/src/apis/verifiedaccess/dist/v1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.verifiedaccess_v1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var verifiedaccess_v1;

(function (verifiedaccess_v1) {
  /**
   * Chrome Verified Access API
   *
   * API for Verified Access chrome extension to provide credential verification for chrome devices connecting to an enterprise network
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const verifiedaccess = google.verifiedaccess('v1');
   * ```
   */
  var Verifiedaccess = function Verifiedaccess(options, google) {
    _classCallCheck(this, Verifiedaccess);

    this.context = {
      _options: options || {},
      google: google
    };
    this.challenge = new Resource$Challenge(this.context);
  };

  verifiedaccess_v1.Verifiedaccess = Verifiedaccess;

  var Resource$Challenge =
  /*#__PURE__*/
  function () {
    function Resource$Challenge(context) {
      _classCallCheck(this, Resource$Challenge);

      this.context = context;
    }

    _createClass(Resource$Challenge, [{
      key: "create",
      value: function create(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://verifiedaccess.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/challenge').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "verify",
      value: function verify(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://verifiedaccess.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/challenge:verify').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Challenge;
  }();

  verifiedaccess_v1.Resource$Challenge = Resource$Challenge;
})(verifiedaccess_v1 = exports.verifiedaccess_v1 || (exports.verifiedaccess_v1 = {}));
