// googleapis/build/src/apis/webfonts/dist/v1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.webfonts_v1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var webfonts_v1;

(function (webfonts_v1) {
  /**
   * Web Fonts Developer API
   *
   * The Google Web Fonts Developer API lets you retrieve information about web fonts served by Google.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const webfonts = google.webfonts('v1');
   * ```
   */
  var Webfonts = function Webfonts(options, google) {
    _classCallCheck(this, Webfonts);

    this.context = {
      _options: options || {},
      google: google
    };
    this.webfonts = new Resource$Webfonts(this.context);
  };

  webfonts_v1.Webfonts = Webfonts;

  var Resource$Webfonts =
  /*#__PURE__*/
  function () {
    function Resource$Webfonts(context) {
      _classCallCheck(this, Resource$Webfonts);

      this.context = context;
    }

    _createClass(Resource$Webfonts, [{
      key: "list",
      value: function list(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://webfonts.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/webfonts').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Webfonts;
  }();

  webfonts_v1.Resource$Webfonts = Resource$Webfonts;
})(webfonts_v1 = exports.webfonts_v1 || (exports.webfonts_v1 = {}));
