// googleapis/build/src/apis/dns/dist/index.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dns_v2beta1 = exports.dns_v1beta2 = exports.dns_v1 = exports.auth = exports.dns = exports.VERSIONS = void 0;
/*! THIS FILE IS AUTO-GENERATED */

var googleapis_common_1 = require("googleapis-common");

var v1_1 = require("./v1");

Object.defineProperty(exports, "dns_v1", {
  enumerable: true,
  get: function get() {
    return v1_1.dns_v1;
  }
});

var v1beta2_1 = require("./v1beta2");

Object.defineProperty(exports, "dns_v1beta2", {
  enumerable: true,
  get: function get() {
    return v1beta2_1.dns_v1beta2;
  }
});

var v2beta1_1 = require("./v2beta1");

Object.defineProperty(exports, "dns_v2beta1", {
  enumerable: true,
  get: function get() {
    return v2beta1_1.dns_v2beta1;
  }
});
exports.VERSIONS = {
  v1: v1_1.dns_v1.Dns,
  v1beta2: v1beta2_1.dns_v1beta2.Dns,
  v2beta1: v2beta1_1.dns_v2beta1.Dns
};

function dns(versionOrOptions) {
  return googleapis_common_1.getAPI('dns', versionOrOptions, exports.VERSIONS, this);
}

exports.dns = dns;
var auth = new googleapis_common_1.AuthPlus();
exports.auth = auth;

var googleapis_common_2 = require("googleapis-common");

Object.defineProperty(exports, "AuthPlus", {
  enumerable: true,
  get: function get() {
    return googleapis_common_2.AuthPlus;
  }
});
