// googleapis/build/src/apis/policytroubleshooter/dist/v1beta.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.policytroubleshooter_v1beta = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var policytroubleshooter_v1beta;

(function (policytroubleshooter_v1beta) {
  /**
   * Policy Troubleshooter API
   *
   *
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const policytroubleshooter = google.policytroubleshooter('v1beta');
   * ```
   */
  var Policytroubleshooter = function Policytroubleshooter(options, google) {
    _classCallCheck(this, Policytroubleshooter);

    this.context = {
      _options: options || {},
      google: google
    };
    this.iam = new Resource$Iam(this.context);
  };

  policytroubleshooter_v1beta.Policytroubleshooter = Policytroubleshooter;

  var Resource$Iam =
  /*#__PURE__*/
  function () {
    function Resource$Iam(context) {
      _classCallCheck(this, Resource$Iam);

      this.context = context;
    }

    _createClass(Resource$Iam, [{
      key: "troubleshoot",
      value: function troubleshoot(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://policytroubleshooter.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1beta/iam:troubleshoot').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Iam;
  }();

  policytroubleshooter_v1beta.Resource$Iam = Resource$Iam;
})(policytroubleshooter_v1beta = exports.policytroubleshooter_v1beta || (exports.policytroubleshooter_v1beta = {}));
