// googleapis/build/src/apis/analyticsreporting/dist/v4.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.analyticsreporting_v4 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var analyticsreporting_v4;

(function (analyticsreporting_v4) {
  /**
   * Analytics Reporting API
   *
   * Accesses Analytics report data.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const analyticsreporting = google.analyticsreporting('v4');
   * ```
   */
  var Analyticsreporting = function Analyticsreporting(options, google) {
    _classCallCheck(this, Analyticsreporting);

    this.context = {
      _options: options || {},
      google: google
    };
    this.reports = new Resource$Reports(this.context);
    this.userActivity = new Resource$Useractivity(this.context);
  };

  analyticsreporting_v4.Analyticsreporting = Analyticsreporting;

  var Resource$Reports =
  /*#__PURE__*/
  function () {
    function Resource$Reports(context) {
      _classCallCheck(this, Resource$Reports);

      this.context = context;
    }

    _createClass(Resource$Reports, [{
      key: "batchGet",
      value: function batchGet(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://analyticsreporting.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v4/reports:batchGet').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Reports;
  }();

  analyticsreporting_v4.Resource$Reports = Resource$Reports;

  var Resource$Useractivity =
  /*#__PURE__*/
  function () {
    function Resource$Useractivity(context) {
      _classCallCheck(this, Resource$Useractivity);

      this.context = context;
    }

    _createClass(Resource$Useractivity, [{
      key: "search",
      value: function search(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://analyticsreporting.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v4/userActivity:search').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Useractivity;
  }();

  analyticsreporting_v4.Resource$Useractivity = Resource$Useractivity;
})(analyticsreporting_v4 = exports.analyticsreporting_v4 || (exports.analyticsreporting_v4 = {}));
