// mongoose/lib/error/dist/cast.dev.js
'use strict';
/*!
 * Module dependencies.
 */

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

var MongooseError = require('./mongooseError');

var get = require('../helpers/get');

var util = require('util');
/**
 * Casting Error constructor.
 *
 * @param {String} type
 * @param {String} value
 * @inherits MongooseError
 * @api private
 */


var CastError =
/*#__PURE__*/
function (_MongooseError) {
  _inherits(CastError, _MongooseError);

  function CastError(type, value, path, reason, schemaType) {
    var _this;

    _classCallCheck(this, CastError);

    // If no args, assume we'll `init()` later.
    if (arguments.length > 0) {
      var stringValue = getStringValue(value);
      var messageFormat = getMessageFormat(schemaType);
      var msg = formatMessage(null, type, stringValue, path, messageFormat);
      _this = _possibleConstructorReturn(this, _getPrototypeOf(CastError).call(this, msg));

      _this.init(type, value, path, reason, schemaType);
    } else {
      _this = _possibleConstructorReturn(this, _getPrototypeOf(CastError).call(this, formatMessage()));
    }

    return _possibleConstructorReturn(_this);
  }
  /*!
   * ignore
   */


  _createClass(CastError, [{
    key: "init",
    value: function init(type, value, path, reason, schemaType) {
      this.stringValue = getStringValue(value);
      this.messageFormat = getMessageFormat(schemaType);
      this.kind = type;
      this.value = value;
      this.path = path;
      this.reason = reason;
    }
    /*!
     * ignore
     * @param {Readonly<CastError>} other
     */

  }, {
    key: "copy",
    value: function copy(other) {
      this.messageFormat = other.messageFormat;
      this.stringValue = other.stringValue;
      this.kind = other.kind;
      this.value = other.value;
      this.path = other.path;
      this.reason = other.reason;
      this.message = other.message;
    }
    /*!
     * ignore
     */

  }, {
    key: "setModel",
    value: function setModel(model) {
      this.model = model;
      this.message = formatMessage(model, this.kind, this.stringValue, this.path, this.messageFormat);
    }
  }]);

  return CastError;
}(MongooseError);

Object.defineProperty(CastError.prototype, 'name', {
  value: 'CastError'
});

function getStringValue(value) {
  var stringValue = util.inspect(value);
  stringValue = stringValue.replace(/^'|'$/g, '"');

  if (!stringValue.startsWith('"')) {
    stringValue = '"' + stringValue + '"';
  }

  return stringValue;
}

function getMessageFormat(schemaType) {
  var messageFormat = get(schemaType, 'options.cast', null);

  if (typeof messageFormat === 'string') {
    return messageFormat;
  }
}
/*!
 * ignore
 */


function formatMessage(model, kind, stringValue, path, messageFormat) {
  if (messageFormat != null) {
    var ret = messageFormat.replace('{KIND}', kind).replace('{VALUE}', stringValue).replace('{PATH}', path);

    if (model != null) {
      ret = ret.replace('{MODEL}', model.modelName);
    }

    return ret;
  } else {
    var _ret = 'Cast to ' + kind + ' failed for value ' + stringValue + ' at path "' + path + '"';

    if (model != null) {
      _ret += ' for model "' + model.modelName + '"';
    }

    return _ret;
  }
}
/*!
 * exports
 */


module.exports = CastError;
