// googleapis/build/src/apis/firebasedynamiclinks/dist/v1.dev.js
"use strict"; // Copyright 2020 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.firebasedynamiclinks_v1 = void 0;
/* eslint-disable @typescript-eslint/no-explicit-any */

/* eslint-disable @typescript-eslint/class-name-casing */

/* eslint-disable @typescript-eslint/no-unused-vars */

/* eslint-disable @typescript-eslint/no-empty-interface */

/* eslint-disable @typescript-eslint/no-namespace */

/* eslint-disable no-irregular-whitespace */

var googleapis_common_1 = require("googleapis-common");

var firebasedynamiclinks_v1;

(function (firebasedynamiclinks_v1) {
  /**
   * Firebase Dynamic Links API
   *
   * Programmatically creates and manages Firebase Dynamic Links.
   *
   * @example
   * ```js
   * const {google} = require('googleapis');
   * const firebasedynamiclinks = google.firebasedynamiclinks('v1');
   * ```
   */
  var Firebasedynamiclinks = function Firebasedynamiclinks(options, google) {
    _classCallCheck(this, Firebasedynamiclinks);

    this.context = {
      _options: options || {},
      google: google
    };
    this.managedShortLinks = new Resource$Managedshortlinks(this.context);
    this.shortLinks = new Resource$Shortlinks(this.context);
    this.v1 = new Resource$V1(this.context);
  };

  firebasedynamiclinks_v1.Firebasedynamiclinks = Firebasedynamiclinks;

  var Resource$Managedshortlinks =
  /*#__PURE__*/
  function () {
    function Resource$Managedshortlinks(context) {
      _classCallCheck(this, Resource$Managedshortlinks);

      this.context = context;
    }

    _createClass(Resource$Managedshortlinks, [{
      key: "create",
      value: function create(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://firebasedynamiclinks.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/managedShortLinks:create').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Managedshortlinks;
  }();

  firebasedynamiclinks_v1.Resource$Managedshortlinks = Resource$Managedshortlinks;

  var Resource$Shortlinks =
  /*#__PURE__*/
  function () {
    function Resource$Shortlinks(context) {
      _classCallCheck(this, Resource$Shortlinks);

      this.context = context;
    }

    _createClass(Resource$Shortlinks, [{
      key: "create",
      value: function create(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://firebasedynamiclinks.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/shortLinks').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$Shortlinks;
  }();

  firebasedynamiclinks_v1.Resource$Shortlinks = Resource$Shortlinks;

  var Resource$V1 =
  /*#__PURE__*/
  function () {
    function Resource$V1(context) {
      _classCallCheck(this, Resource$V1);

      this.context = context;
    }

    _createClass(Resource$V1, [{
      key: "getLinkStats",
      value: function getLinkStats(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://firebasedynamiclinks.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/{dynamicLink}/linkStats').replace(/([^:]\/)\/+/g, '$1'),
            method: 'GET'
          }, options),
          params: params,
          requiredParams: ['dynamicLink'],
          pathParams: ['dynamicLink'],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "installAttribution",
      value: function installAttribution(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://firebasedynamiclinks.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/installAttribution').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }, {
      key: "reopenAttribution",
      value: function reopenAttribution(paramsOrCallback, optionsOrCallback, callback) {
        var params = paramsOrCallback || {};
        var options = optionsOrCallback || {};

        if (typeof paramsOrCallback === 'function') {
          callback = paramsOrCallback;
          params = {};
          options = {};
        }

        if (typeof optionsOrCallback === 'function') {
          callback = optionsOrCallback;
          options = {};
        }

        var rootUrl = options.rootUrl || 'https://firebasedynamiclinks.googleapis.com/';
        var parameters = {
          options: Object.assign({
            url: (rootUrl + '/v1/reopenAttribution').replace(/([^:]\/)\/+/g, '$1'),
            method: 'POST'
          }, options),
          params: params,
          requiredParams: [],
          pathParams: [],
          context: this.context
        };

        if (callback) {
          googleapis_common_1.createAPIRequest(parameters, callback);
        } else {
          return googleapis_common_1.createAPIRequest(parameters);
        }
      }
    }]);

    return Resource$V1;
  }();

  firebasedynamiclinks_v1.Resource$V1 = Resource$V1;
})(firebasedynamiclinks_v1 = exports.firebasedynamiclinks_v1 || (exports.firebasedynamiclinks_v1 = {}));
