// kareem/dist/index.dev.js
'use strict';

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

function Kareem() {
  this._pres = new Map();
  this._posts = new Map();
}

Kareem.prototype.execPre = function (name, context, args, callback) {
  if (arguments.length === 3) {
    callback = args;
    args = [];
  }

  var pres = get(this._pres, name, []);
  var numPres = pres.length;
  var numAsyncPres = pres.numAsync || 0;
  var currentPre = 0;
  var asyncPresLeft = numAsyncPres;
  var done = false;
  var $args = args;

  if (!numPres) {
    return process.nextTick(function () {
      callback(null);
    });
  }

  var next = function next() {
    if (currentPre >= numPres) {
      return;
    }

    var pre = pres[currentPre];

    if (pre.isAsync) {
      var args = [decorateNextFn(_next), decorateNextFn(function (error) {
        if (error) {
          if (done) {
            return;
          }

          done = true;
          return callback(error);
        }

        if (--asyncPresLeft === 0 && currentPre >= numPres) {
          return callback(null);
        }
      })];
      callMiddlewareFunction(pre.fn, context, args, args[0]);
    } else if (pre.fn.length > 0) {
      var args = [decorateNextFn(_next)];

      var _args = arguments.length >= 2 ? arguments : [null].concat($args);

      for (var i = 1; i < _args.length; ++i) {
        args.push(_args[i]);
      }

      callMiddlewareFunction(pre.fn, context, args, args[0]);
    } else {
      var maybePromise = null;

      try {
        maybePromise = pre.fn.call(context);
      } catch (err) {
        if (err != null) {
          return callback(err);
        }
      }

      if (isPromise(maybePromise)) {
        maybePromise.then(function () {
          return _next();
        }, function (err) {
          return _next(err);
        });
      } else {
        if (++currentPre >= numPres) {
          if (asyncPresLeft > 0) {
            // Leave parallel hooks to run
            return;
          } else {
            return process.nextTick(function () {
              callback(null);
            });
          }
        }

        next();
      }
    }
  };

  next.apply(null, [null].concat(args));

  function _next(error) {
    if (error) {
      if (done) {
        return;
      }

      done = true;
      return callback(error);
    }

    if (++currentPre >= numPres) {
      if (asyncPresLeft > 0) {
        // Leave parallel hooks to run
        return;
      } else {
        return callback(null);
      }
    }

    next.apply(context, arguments);
  }
};

Kareem.prototype.execPreSync = function (name, context, args) {
  var pres = get(this._pres, name, []);
  var numPres = pres.length;

  for (var i = 0; i < numPres; ++i) {
    pres[i].fn.apply(context, args || []);
  }
};

Kareem.prototype.execPost = function (name, context, args, options, callback) {
  if (arguments.length < 5) {
    callback = options;
    options = null;
  }

  var posts = get(this._posts, name, []);
  var numPosts = posts.length;
  var currentPost = 0;
  var firstError = null;

  if (options && options.error) {
    firstError = options.error;
  }

  if (!numPosts) {
    return process.nextTick(function () {
      callback.apply(null, [firstError].concat(args));
    });
  }

  var next = function next() {
    var post = posts[currentPost].fn;
    var numArgs = 0;
    var argLength = args.length;
    var newArgs = [];

    for (var i = 0; i < argLength; ++i) {
      numArgs += args[i] && args[i]._kareemIgnore ? 0 : 1;

      if (!args[i] || !args[i]._kareemIgnore) {
        newArgs.push(args[i]);
      }
    }

    if (firstError) {
      if (post.length === numArgs + 2) {
        var _cb = decorateNextFn(function (error) {
          if (error) {
            firstError = error;
          }

          if (++currentPost >= numPosts) {
            return callback.call(null, firstError);
          }

          next();
        });

        callMiddlewareFunction(post, context, [firstError].concat(newArgs).concat([_cb]), _cb);
      } else {
        if (++currentPost >= numPosts) {
          return callback.call(null, firstError);
        }

        next();
      }
    } else {
      var _cb2 = decorateNextFn(function (error) {
        if (error) {
          firstError = error;
          return next();
        }

        if (++currentPost >= numPosts) {
          return callback.apply(null, [null].concat(args));
        }

        next();
      });

      if (post.length === numArgs + 2) {
        // Skip error handlers if no error
        if (++currentPost >= numPosts) {
          return callback.apply(null, [null].concat(args));
        }

        return next();
      }

      if (post.length === numArgs + 1) {
        callMiddlewareFunction(post, context, newArgs.concat([_cb2]), _cb2);
      } else {
        var error;
        var maybePromise;

        try {
          maybePromise = post.apply(context, newArgs);
        } catch (err) {
          error = err;
          firstError = err;
        }

        if (isPromise(maybePromise)) {
          return maybePromise.then(function () {
            return _cb2();
          }, function (err) {
            return _cb2(err);
          });
        }

        if (++currentPost >= numPosts) {
          return callback.apply(null, [error].concat(args));
        }

        next(error);
      }
    }
  };

  next();
};

Kareem.prototype.execPostSync = function (name, context, args) {
  var posts = get(this._posts, name, []);
  var numPosts = posts.length;

  for (var i = 0; i < numPosts; ++i) {
    posts[i].fn.apply(context, args || []);
  }
};

Kareem.prototype.createWrapperSync = function (name, fn) {
  var kareem = this;
  return function syncWrapper() {
    kareem.execPreSync(name, this, arguments);
    var toReturn = fn.apply(this, arguments);
    kareem.execPostSync(name, this, [toReturn]);
    return toReturn;
  };
};

function _handleWrapError(instance, error, name, context, args, options, callback) {
  if (options.useErrorHandlers) {
    var _options = {
      error: error
    };
    return instance.execPost(name, context, args, _options, function (error) {
      return typeof callback === 'function' && callback(error);
    });
  } else {
    return typeof callback === 'function' ? callback(error) : undefined;
  }
}

Kareem.prototype.wrap = function (name, fn, context, args, options) {
  var lastArg = args.length > 0 ? args[args.length - 1] : null;
  var argsWithoutCb = typeof lastArg === 'function' ? args.slice(0, args.length - 1) : args;

  var _this = this;

  options = options || {};
  var checkForPromise = options.checkForPromise;
  this.execPre(name, context, args, function (error) {
    if (error) {
      var numCallbackParams = options.numCallbackParams || 0;
      var errorArgs = options.contextParameter ? [context] : [];

      for (var i = errorArgs.length; i < numCallbackParams; ++i) {
        errorArgs.push(null);
      }

      return _handleWrapError(_this, error, name, context, errorArgs, options, lastArg);
    }

    var end = typeof lastArg === 'function' ? args.length - 1 : args.length;
    var numParameters = fn.length;
    var ret = fn.apply(context, args.slice(0, end).concat(_cb));

    if (checkForPromise) {
      if (ret != null && typeof ret.then === 'function') {
        // Thenable, use it
        return ret.then(function (res) {
          return _cb(null, res);
        }, function (err) {
          return _cb(err);
        });
      } // If `fn()` doesn't have a callback argument and doesn't return a
      // promise, assume it is sync


      if (numParameters < end + 1) {
        return _cb(null, ret);
      }
    }

    function _cb() {
      var args = arguments;
      var argsWithoutError = Array.prototype.slice.call(arguments, 1);

      if (options.nullResultByDefault && argsWithoutError.length === 0) {
        argsWithoutError.push(null);
      }

      if (arguments[0]) {
        // Assume error
        return _handleWrapError(_this, arguments[0], name, context, argsWithoutError, options, lastArg);
      } else {
        _this.execPost(name, context, argsWithoutError, function () {
          if (arguments[0]) {
            return typeof lastArg === 'function' ? lastArg(arguments[0]) : undefined;
          }

          return typeof lastArg === 'function' ? lastArg.apply(context, arguments) : undefined;
        });
      }
    }
  });
};

Kareem.prototype.filter = function (fn) {
  var _this2 = this;

  var clone = this.clone();
  var pres = Array.from(clone._pres.keys());

  var _loop = function _loop() {
    var name = _pres[_i];

    var hooks = _this2._pres.get(name).map(function (h) {
      return Object.assign({}, h, {
        name: name
      });
    }).filter(fn);

    if (hooks.length === 0) {
      clone._pres["delete"](name);

      return "continue";
    }

    hooks.numAsync = hooks.filter(function (h) {
      return h.isAsync;
    }).length;

    clone._pres.set(name, hooks);
  };

  for (var _i = 0, _pres = pres; _i < _pres.length; _i++) {
    var _ret = _loop();

    if (_ret === "continue") continue;
  }

  var posts = Array.from(clone._posts.keys());

  var _loop2 = function _loop2() {
    var name = _posts[_i2];

    var hooks = _this2._posts.get(name).map(function (h) {
      return Object.assign({}, h, {
        name: name
      });
    }).filter(fn);

    if (hooks.length === 0) {
      clone._posts["delete"](name);

      return "continue";
    }

    clone._posts.set(name, hooks);
  };

  for (var _i2 = 0, _posts = posts; _i2 < _posts.length; _i2++) {
    var _ret2 = _loop2();

    if (_ret2 === "continue") continue;
  }

  return clone;
};

Kareem.prototype.hasHooks = function (name) {
  return this._pres.has(name) || this._posts.has(name);
};

Kareem.prototype.createWrapper = function (name, fn, context, options) {
  var _this = this;

  if (!this.hasHooks(name)) {
    // Fast path: if there's no hooks for this function, just return the
    // function wrapped in a nextTick()
    return function () {
      var _arguments = arguments,
          _this3 = this;

      process.nextTick(function () {
        return fn.apply(_this3, _arguments);
      });
    };
  }

  return function () {
    var _context = context || this;

    var args = Array.prototype.slice.call(arguments);

    _this.wrap(name, fn, _context, args, options);
  };
};

Kareem.prototype.pre = function (name, isAsync, fn, error, unshift) {
  var options = {};

  if (_typeof(isAsync) === 'object' && isAsync != null) {
    options = isAsync;
    isAsync = options.isAsync;
  } else if (typeof arguments[1] !== 'boolean') {
    error = fn;
    fn = isAsync;
    isAsync = false;
  }

  var pres = get(this._pres, name, []);

  this._pres.set(name, pres);

  if (isAsync) {
    pres.numAsync = pres.numAsync || 0;
    ++pres.numAsync;
  }

  if (typeof fn !== 'function') {
    throw new Error('pre() requires a function, got "' + _typeof(fn) + '"');
  }

  if (unshift) {
    pres.unshift(Object.assign({}, options, {
      fn: fn,
      isAsync: isAsync
    }));
  } else {
    pres.push(Object.assign({}, options, {
      fn: fn,
      isAsync: isAsync
    }));
  }

  return this;
};

Kareem.prototype.post = function (name, options, fn, unshift) {
  var hooks = get(this._posts, name, []);

  if (typeof options === 'function') {
    unshift = !!fn;
    fn = options;
    options = {};
  }

  if (typeof fn !== 'function') {
    throw new Error('post() requires a function, got "' + _typeof(fn) + '"');
  }

  if (unshift) {
    hooks.unshift(Object.assign({}, options, {
      fn: fn
    }));
  } else {
    hooks.push(Object.assign({}, options, {
      fn: fn
    }));
  }

  this._posts.set(name, hooks);

  return this;
};

Kareem.prototype.clone = function () {
  var n = new Kareem();
  var _iteratorNormalCompletion = true;
  var _didIteratorError = false;
  var _iteratorError = undefined;

  try {
    for (var _iterator = this._pres.keys()[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
      var key = _step.value;

      var clone = this._pres.get(key).slice();

      clone.numAsync = this._pres.get(key).numAsync;

      n._pres.set(key, clone);
    }
  } catch (err) {
    _didIteratorError = true;
    _iteratorError = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion && _iterator["return"] != null) {
        _iterator["return"]();
      }
    } finally {
      if (_didIteratorError) {
        throw _iteratorError;
      }
    }
  }

  var _iteratorNormalCompletion2 = true;
  var _didIteratorError2 = false;
  var _iteratorError2 = undefined;

  try {
    for (var _iterator2 = this._posts.keys()[Symbol.iterator](), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {
      var _key = _step2.value;

      n._posts.set(_key, this._posts.get(_key).slice());
    }
  } catch (err) {
    _didIteratorError2 = true;
    _iteratorError2 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion2 && _iterator2["return"] != null) {
        _iterator2["return"]();
      }
    } finally {
      if (_didIteratorError2) {
        throw _iteratorError2;
      }
    }
  }

  return n;
};

Kareem.prototype.merge = function (other, clone) {
  clone = arguments.length === 1 ? true : clone;
  var ret = clone ? this.clone() : this;
  var _iteratorNormalCompletion3 = true;
  var _didIteratorError3 = false;
  var _iteratorError3 = undefined;

  try {
    var _loop3 = function _loop3() {
      var key = _step3.value;
      var sourcePres = get(ret._pres, key, []);

      var deduplicated = other._pres.get(key). // Deduplicate based on `fn`
      filter(function (p) {
        return sourcePres.map(function (_p) {
          return _p.fn;
        }).indexOf(p.fn) === -1;
      });

      var combined = sourcePres.concat(deduplicated);
      combined.numAsync = sourcePres.numAsync || 0;
      combined.numAsync += deduplicated.filter(function (p) {
        return p.isAsync;
      }).length;

      ret._pres.set(key, combined);
    };

    for (var _iterator3 = other._pres.keys()[Symbol.iterator](), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {
      _loop3();
    }
  } catch (err) {
    _didIteratorError3 = true;
    _iteratorError3 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion3 && _iterator3["return"] != null) {
        _iterator3["return"]();
      }
    } finally {
      if (_didIteratorError3) {
        throw _iteratorError3;
      }
    }
  }

  var _iteratorNormalCompletion4 = true;
  var _didIteratorError4 = false;
  var _iteratorError4 = undefined;

  try {
    var _loop4 = function _loop4() {
      var key = _step4.value;
      var sourcePosts = get(ret._posts, key, []);

      var deduplicated = other._posts.get(key).filter(function (p) {
        return sourcePosts.indexOf(p) === -1;
      });

      ret._posts.set(key, sourcePosts.concat(deduplicated));
    };

    for (var _iterator4 = other._posts.keys()[Symbol.iterator](), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {
      _loop4();
    }
  } catch (err) {
    _didIteratorError4 = true;
    _iteratorError4 = err;
  } finally {
    try {
      if (!_iteratorNormalCompletion4 && _iterator4["return"] != null) {
        _iterator4["return"]();
      }
    } finally {
      if (_didIteratorError4) {
        throw _iteratorError4;
      }
    }
  }

  return ret;
};

function get(map, key, def) {
  if (map.has(key)) {
    return map.get(key);
  }

  return def;
}

function callMiddlewareFunction(fn, context, args, next) {
  var maybePromise;

  try {
    maybePromise = fn.apply(context, args);
  } catch (error) {
    return next(error);
  }

  if (isPromise(maybePromise)) {
    maybePromise.then(function () {
      return next();
    }, function (err) {
      return next(err);
    });
  }
}

function isPromise(v) {
  return v != null && typeof v.then === 'function';
}

function decorateNextFn(fn) {
  var called = false;

  var _this = this;

  return function () {
    var _arguments2 = arguments;

    // Ensure this function can only be called once
    if (called) {
      return;
    }

    called = true; // Make sure to clear the stack so try/catch doesn't catch errors
    // in subsequent middleware

    return process.nextTick(function () {
      return fn.apply(_this, _arguments2);
    });
  };
}

module.exports = Kareem;
